prow/crier/crier.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package crier sets GitHub statuses and writes comments for test statuses.
+package crier
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	commentTag  = "<!-- test report -->"
+	guberPrefix = "https://k8s-gubernator.appspot.com/pr/"
+)
+
+type Report struct {
+	RepoOwner string `json:"repo_owner"`
+	RepoName  string `json:"repo_name"`
+	Author    string `json:"author"`
+	Number    int    `json:"number"`
+	Commit    string `json:"commit"`
+
+	Context      string `json:"context"`
+	State        string `json:"state"`
+	Description  string `json:"description"`
+	RerunCommand string `json:"rerun_command"`
+	URL          string `json:"url"`
+}
+
+type Server struct {
+	rc     chan Report
+	ghc    GitHubClient
+	notify chan struct{}
+}
+
+type GitHubClient interface {
+	CreateStatus(org, repo, ref string, s github.Status) error
+	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error)
+	CreateComment(org, repo string, number int, comment string) error
+	DeleteComment(org, repo string, ID int) error
+}
+
+func NewServer(ghc GitHubClient) *Server {
+	return &Server{
+		rc:  make(chan Report),
+		ghc: ghc,
+	}
+}
+
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
```

Comment:
You should restrict the URL here too, so it's easier to extend this in the future.